### PR TITLE
Failing test for #4650

### PR DESCRIPTION
**Pull request by ruudk**
Trying to fix https://github.com/phpstan/phpstan/issues/4650

Let's see if this one test fails, and then try to fix it.

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
You can verify that locally with `vendor/bin/phing tests-fast`, no need to go through CI each time :) See the README for more commands like that.

**Comment by ruudk** [comment]
I was trying to test it locally but a lot of the tests failed for some reason. Will try `vendor/bin/phing tests-fast`. thanks.

**Comment by ondrejmirtes** [comment]
That also shouldn't happen :) It's all green on my machine.

**Comment by ruudk** [comment]
Interesting, doing another `composer install` solved it. Green now. But it should fail :D

File: src/Analyser/MutatingScope.php [hunk 1]
```diff
@@ -754,8 +754,13 @@ private function resolveType(Expr $node): Type
 		}
 
 		if ($node instanceof Expr\BinaryOp\Identical) {
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:756]: Also do it for `NotIdentical` :)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/MutatingScope.php (hunk 1 above)
Comment (ruudk) [reply src/Analyser/MutatingScope.php:756]: Done

**Comment by ruudk** [comment]
I feel that this needs more tests to make sure it works as expected. I want to add them, but don't know where.

**Comment by ondrejmirtes** [comment]
1) Test also the NotIdentical expression (`!==`)
2) Test `assertType` for the same expressions in the same test.

That would be sufficient from my perspective :)

File: tests/PHPStan/Analyser/data/bug-4650.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+namespace Bug4213;
+
+use function PHPStan\Analyser\assertNativeType;
+use function PHPStan\Analyser\assertType;
+
+/**
+ * @phpstan-param non-empty-array<string|int> $idx
+ */
+function (array $idx): void {
+	assertType('bool', $idx === []);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-4650.php:12]: This should be false obviously

File: tests/PHPStan/Analyser/data/bug-4650.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	assertNativeType('bool', $idx === []);
+	assertType('bool', $idx !== []);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-4650.php:14]: This should be true

Comment (ruudk) [reply tests/PHPStan/Analyser/data/bug-4650.php:14]: Then it fails
```
1) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "/Users/Ruud/opensource/phpstan-src/tests/PHPStan/Analyser/data/bug-4650.php:12" ('type', '/Users/Ruud/opensource/phpsta...50.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\BooleanType Object (), 12)
Expected type false, got type bool in /Users/Ruud/opensource/phpstan-src/tests/PHPStan/Analyser/data/bug-4650.php on line 12.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'false'
+'bool'

/Users/Ruud/opensource/phpstan-src/tests/PHPStan/Analyser/NodeScopeResolverTest.php:11333

2) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "/Users/Ruud/opensource/phpstan-src/tests/PHPStan/Analyser/data/bug-4650.php:14" ('type', '/Users/Ruud/opensource/phpsta...50.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\BooleanType Object (), 14)
Expected type true, got type bool in /Users/Ruud/opensource/phpstan-src/tests/PHPStan/Analyser/data/bug-4650.php on line 14.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'true'
+'bool'
```

It's also why I think that the test is not good enough. Shouldn't it be tested with `treatPhpDocTypesAsCertain: false`?

**Comment by ruudk** [comment]
@ondrejmirtes Any advice for this PR?

**Comment by ondrejmirtes** [comment]
I see it now - the problem is that the tested function is an anonymous function, not a named one. It works as expected in a function: https://phpstan.org/r/29929e67-5641-4a4b-8c3d-d58d65bb58d8

PHPDocs above closures are not yet supported.

It's best to structure the test in a class + method rather than function as classes can be autoloaded and you don't have to `require_once` the file manually. Thanks.

**Comment by VincentLanglet** [comment]
Hi @ruudk I'm really interested by this PR, do you have some time to finish it ? :)

**Comment by ruudk** [comment]
I'll try to have another look at it this week.

**Comment by ruudk** [comment]
@VincentLanglet Sorry, I won't be fixing this anytime soon.

**Comment by VincentLanglet** [comment]
> @VincentLanglet Sorry, I won't be fixing this anytime soon.

Will take a look

**Comment by ondrejmirtes** [comment]
I've modernized the test so that it looks how tests are currently written, but it still fails.

**Comment by ondrejmirtes** [comment]
Alright, the fix is correct, but the test is failing of unrelated reasons.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Now it works :) Thanks everyone.

**Comment by ruudk** [comment]
Thanks 🙌

**Pull request merged**